src/main/kotlin/org/rust/lang/core/types/RustEnumType.kt

```diff
@@ -4,12 +4,21 @@ import com.intellij.codeInsight.completion.CompletionUtil
 import org.rust.lang.core.psi.RustEnumItemElement
 import org.rust.lang.core.types.visitors.RustTypeVisitor
 
-class RustEnumType(enum: RustEnumItemElement) : RustStructOrEnumTypeBase() {
+class RustEnumType(
+    enum: RustEnumItemElement,
+    typeArguments: List<RustType> = emptyList()
+) : RustStructOrEnumTypeBase(typeArguments) {
```

Comment:
Yep, the decorator approach does not work here at all because sometime we want to know the precise type of the thing ([example](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt#L122)).

Also, it would make type visitors massively more complicated, because you'll need to deal with possibility of `TypeWithSubstitution(TypeWithSubstitution(TypeWithSubstitution(BaseType)))`.